async_engine.cpp

```diff
@@ -0,0 +1,92 @@
+#include "async_engine.hpp"
+#include "socket_exception.hpp"
+#include <iostream>
+#include <sys/select.h>
+
+AsyncEngine::AsyncEngine(Socket_sptr master_socket) : m_master_socket(master_socket)
+{
+    FD_ZERO(&m_fd_set_read);
+    FD_SET(*master_socket, &m_fd_set_read);
+    m_max_fd = *master_socket;
+}
+
+AsyncEngine::~AsyncEngine()
+{
+    try
+    {
+        m_master_socket.reset();
+    }
+    catch (SocketException & se)
```

Comment:
Such a nice solution. :)
But you still have a vector of other sockets will be cleaned up on engine destruction - what about those exceptions?
I fact, we need to decide: if socket closing error is critical for us, it should crash application, so we should not try to catch it. Otherwise, this exception should be processed in `Socket` destructor code (just suppressed i guess).